internal/webhook/external/github.go

```diff
@@ -0,0 +1,147 @@
+package external
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+
+	gh "github.com/google/go-github/v71/github"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	xhttp "github.com/akuity/kargo/internal/http"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+// githubHandler handles push events for github.
+// After the request has been authenticated,
+// the kubeclient is queried for all warehouses that contain a subscription
+// to the repo in question. Those warehouses are then patched with a special
+// annotation that signals down stream logic to refresh the warehouse.
+func githubHandler(c client.Client) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+		logger.Debug("identifying source repository")
+
+		secret, ok := os.LookupEnv("GH_WEBHOOK_SECRET")
+		if !ok {
+			logger.Error(
+				errors.New("GH_WEBHOOK_SECRET unset"),
+				"environment misconfiguration",
+			)
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					// keep it vague, no need to leak config details
+					errors.New("unexpected server error"),
```

Comment:
No need for this sort of thing when [this](https://github.com/akuity/kargo/pull/4000/files#r2082603066) is fixed.